main/lsp/ErrorReporter.cc

```diff
@@ -0,0 +1,98 @@
+#include "main/lsp/ErrorReporter.h"
+#include "core/lsp/TypecheckEpochManager.h"
+#include "main/lsp/LSPMessage.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+
+namespace sorbet::realmain::lsp {
+using namespace std;
+ErrorReporter::ErrorReporter(shared_ptr<const LSPConfiguration> config) : config(move(config)) {}
+
+// Used for unit tests
+const std::vector<ErrorStatus> &ErrorReporter::getFileErrorStatuses() const {
+    return fileErrorStatuses;
+};
+
+vector<core::FileRef> ErrorReporter::filesUpdatedSince(u4 epoch) {
+    vector<core::FileRef> filesUpdatedSince;
+    for (size_t i = 1; i < fileErrorStatuses.size(); ++i) {
+        if (fileErrorStatuses[i].sentEpoch >= epoch) {
+            filesUpdatedSince.push_back(core::FileRef(i));
+        }
+    }
+    return filesUpdatedSince;
+}
+
+void ErrorReporter::pushDiagnostics(u4 epoch, core::FileRef file, const vector<unique_ptr<core::Error>> &errors,
```

Comment:
I thought of another bug.

Starting point: `foo.rb` defines a method, `bar.rb` uses the method. Neither have errors.

```ruby
// foo.rb
module Foo
  extend T::Sig
  sig {returns(Integer)}
  def self.foo
    1
  end
end

// bar.rb
module Bar
  extend T::Sig
  sig {returns(Integer)}
  def self.bar
    Foo.foo
  end
end
```

User updates `foo.rb` to add a second method _and_ to change the return value of `Foo.foo` to `String`. It takes the slow path. Outcome:`bar.rb` has errors (because the return value doesn't match the expected type), `foo.rb` does not.

A fast path updates `foo.rb` so that `Foo.foo` returns `Integer`, resolving the error in `bar.rb` from the slow path.

If the fast path update preempts the slow path and is reported before the slow path update, `sentEpoch` for `bar.rb` will not be updated since `errors.empty() && fileErrorStatus.hasErrors == false`. When the slow path goes to report an error for `bar.rb`, it _will_ be reported since 1) `bar.rb`'s epoch is still < `epoch` since it has not been edited, 2) `bar.rb`'s sentEpoch is still < epoch, and 3) `!errors.empty()`.

Reply:
Here's one idea on how to solve this problem:

* Rename `sentEpoch` to `lastReportedEpoch`.
* Update `lastReportedEpoch` if `!(fileErrorStatus.lastReportedEpoch > epoch)`

With this design, if `lastReportedEpoch >= epoch`, then we know that that file was typechecked during and after epoch `epoch`. In other words, the client's state for that given file matches that of epoch `epoch`.

One consequence of the above change is that `filesUpdatedSince` will now contain ~every file in the project if a slow path is canceled near its end (since it reports on all files), which is no good. I think it's safe to rename `filesUpdatedSince` to `filesWithErrorsSince` and to report only files that currently have errors. If a file has no errors and isn't in the set of files to be typechecked given a set of edits, then it should be safe to not retypecheck it.

Reply:
If we make this change, we should be able to discard the file.data(gs).epoch check at the start of `pushDiagnostics`, since if `file.data(gs).epoch > epoch` then it must be true that `fileErrorStatus.sentEpoch > epoch`.